faunadb-java-dsl/src/main/java/com/faunadb/client/types/Codec.java

```diff
@@ -66,86 +83,131 @@
+  Codec<Double> DOUBLE = Cast.mapTo(DoubleV.class, Cast.<DoubleV, Double>scalarValue(), Cast.DOUBLE_TO_VALUE);
+
+  /**
+   * Coerce a {@link Value} to a {@link Float}
+   */
+  Codec<Float> FLOAT = Cast.map(DOUBLE, Cast.DOUBLE_TO_FLOAT, Cast.FLOAT_TO_DOUBLE);
 
   /**
    * Coerce a {@link Value} to a {@link Boolean}
    */
-  Codec<Boolean> BOOLEAN = Cast.mapTo(BooleanV.class, Cast.<BooleanV, Boolean>scalarValue());
+  Codec<Boolean> BOOLEAN = Cast.mapTo(BooleanV.class, Cast.<BooleanV, Boolean>scalarValue(), Cast.BOOLEAN_TO_VALUE);
 
   /**
    * Coerce a {@link Value} to a {@link LocalDate}
    */
-  Codec<LocalDate> DATE = Cast.mapTo(DateV.class, Cast.<DateV, LocalDate>scalarValue());
+  Codec<LocalDate> DATE = Cast.mapTo(DateV.class, Cast.<DateV, LocalDate>scalarValue(), Cast.LOCAL_DATE_TO_VALUE);
 
   /**
    * Coerce a {@link Value} to an {@link ImmutableList} of {@link Value}
    */
-  Codec<ImmutableList<Value>> ARRAY = Cast.mapTo(ArrayV.class, new Function<ArrayV, ImmutableList<Value>>() {
-    @Override
-    public ImmutableList<Value> apply(ArrayV input) {
-      return input.values;
-    }
-  });
+  Codec<ImmutableList<Value>> ARRAY = Cast.mapTo(ArrayV.class, Cast.VALUE_TO_LIST, Cast.LIST_TO_VALUE);
 
   /**
    * Coerce a {@link Value} to an {@link ImmutableMap} of {@link Value}
    */
-  Codec<ImmutableMap<String, Value>> OBJECT = Cast.mapTo(ObjectV.class, new Function<ObjectV, ImmutableMap<String, Value>>() {
-    @Override
-    public ImmutableMap<String, Value> apply(ObjectV input) {
-      return input.values;
-    }
-  });
+  Codec<ImmutableMap<String, Value>> OBJECT = Cast.mapTo(ObjectV.class, Cast.VALUE_TO_MAP, Cast.MAP_TO_VALUE);
 
   /**
    * Coerce a {@link Value} to an array of bytes
    */
-  Codec<byte[]> BYTES = Cast.mapTo(BytesV.class, Cast.<BytesV, byte[]>scalarValue());
+  Codec<byte[]> BYTES = Cast.mapTo(BytesV.class, Cast.<BytesV, byte[]>scalarValue(), Cast.BYTES_TO_VALUE);
 }
 
 final class Cast {
 
-  static <V extends Value, O> Codec<O> mapTo(final Class<V> clazz, final Function<V, O> extractValue) {
+  static <V extends Value, O> Codec<O> mapTo(final Class<V> clazz, final Function<V, O> extractValue, final Function<O, Value> wrapValue) {
     return new Codec<O>() {
       @Override
-      public Result<O> apply(Value input) {
-        return cast(clazz, input).map(extractValue);
+      public Result<O> decode(Value input) {
+        try {
+          return cast(clazz, input).map(extractValue);
+        } catch (Exception ex) {
+          return Result.fail(ex.getMessage(), ex);
+        }
+      }
+
+      @Override
+      public Result<Value> encode(O value) {
+        try {
+          return Result.success(wrapValue.apply(value));
+        } catch (Exception ex) {
+          return Result.fail(ex.getMessage(), ex);
+        }
+      }
+    };
+  }
+
+  static <I, O> Codec<O> map(final Codec<I> codec, final Function<I, O> mapFunction, final Function<O, I> unmapFunction) {
```

Comment:
I like this function but, usually the expected signature of a map function is `<A, B> B map(A value, Function<A, B> fn);`. I'd rename it to something else. Like `mapWith`, `mapC`... idk.